src/safeposix/syscalls/fs_calls.rs

```diff
@@ -988,118 +1036,122 @@ impl Cage {
-                        },
-                        Inode::Dir(ref mut dir_inode_obj) => {
-                            dir_inode_obj.refcount -= 1;
+                        }
 
-                            //if there are still references to the file, then do nothing to the file
-                            if dir_inode_obj.refcount != 0 {
-                                return 0;
-                            }
-                            if fobjtable.contains_key(&inodenum) {
-                                //close and delete the file in the object table
-                                fobjtable.get(&inodenum).unwrap().close().unwrap();
-                                fobjtable.remove(&inodenum);
-                            }
-                            if dir_inode_obj.refcount == 0 {
-                                if dir_inode_obj.linkcount == 0  {
-                                    mutmetadata.inodetable.remove(&inodenum);
-                                    let sysfilename = format!("{}{}", FILEDATAPREFIX, inodenum);
-                                    interface::removefile(sysfilename).unwrap();
-            
-                                } 
-                            }
-                        },
-                        Inode::CharDev(ref mut char_inode_obj) => {
-                            char_inode_obj.refcount -= 1;
+                        fobjtable.get(&inodenum).unwrap().close().unwrap();
+                        fobjtable.remove(&inodenum);
 
-                            //if there are still references to the file, then do nothing to the file
-                            if char_inode_obj.refcount != 0 {
-                                return 0;
+                        //if there are still references to the file, then do nothing to the file
+                        if normalfile_inode_obj.refcount != 0 {
+                            return 0;
+                        }
+                        
+                        if normalfile_inode_obj.linkcount == 0  {
+                            
+                            //removing the file from the entire filesystem
+                            mutmetadata.inodetable.remove(&inodenum);
+                            let sysfilename = format!("{}{}", FILEDATAPREFIX, inodenum);
+                            interface::removefile(sysfilename).unwrap();
+    
+                        } 
+                    },
+                    Inode::Dir(ref mut dir_inode_obj) => {
+                        dir_inode_obj.refcount -= 1;
+
+                        //if it's not a reg file, then we have nothing to close
+                        if !is_reg(dir_inode_obj.mode) {
+                            match fobjtable.get(&inodenum) {
+                                Some(_) => {syscall_error(Errno::ENOEXEC, "close or dup", "Non-regular file in file object table");},
+                                None => {return 0;}
                             }
+                        }
```

Comment:
No dirs are regular files, so rather than this whole statement, as mentioned earlier, don't return, just return a syscall error if there is a corresponding entry in the fobj table